### ISSUE TITLE
Role-related fields on User struct are confusing and overly restrictive when using MongoDB 2.6+

### ISSUE THREAD

**Issue by mhurne** [go-mgo/mgo#236]
The role-related fields on the `User` struct are confusing and overly restrictive when using MongoDB 2.6+. I suspect this is an artifact of backward compatibility. As-is, it is not possible to use a `User` and `UpsertUser` to create a user in some database other than the `admin` database with roles on other databases. For example, this is perfectly valid in MongoDB 3.0, but can't be achieved with `User` and `UpsertUser`:

```
> use foo
switched to db foo
> db.createUser({ user: "johndoe", pwd: "...", roles: [ { role: "read", db: "bar" } ] })
Successfully added user: {
	"user" : "johndoe",
	"roles" : [
		{
			"role" : "read",
			"db" : "bar"
		}
	]
}
> use admin
switched to db admin
> db.system.users.find({ user: "johndoe" }).pretty()
{
	"_id" : "foo.johndoe",
	"user" : "johndoe",
	"db" : "foo",
	"credentials" : {
		"SCRAM-SHA-1" : {
			...
		}
	},
	"roles" : [
		{
			"role" : "read",
			"db" : "bar"
		}
	]
}
```

Initially, I thought that `Roles` was a convenience to assign a role in the database the user is created in while `OtherDBRoles` was the way to assign a role in some other database. However, `UpsertUser` rejects an attempt to create a user in a database other than `admin` if `OtherDBRoles` is populated. Upon further investigation, I found that `OtherDBRoles` was a property of users in Mongo 2.4 but was dropped in 2.6. I also see that mgo was modified to put the contents of `OtherDBRoles` in the user's `roles` property instead of `OtherDBRoles` when the Mongo server is 2.6 or later. But the restriction that `OtherDBRoles` must only be used when creating the user in the `admin` database was never lifted, regardless of the Mongo server version.

At minimum, can we somehow modify mgo to allow creating a user in a database other than `admin` with `OtherDBRoles` populated if the Mongo server is 2.6 or later?

At best, can we somehow drop support for 2.4 and clean up `User` to no longer have an `OtherDBRoles` field whatsoever?

This is causing me some pain in my effort to tackle https://github.com/hashicorp/vault/issues/362

I recognize that I could drop down to calling `Database.Run` myself, but would prefer to avoid that if possible.

**Comment by niemeyer** [comment]
Thanks for reaching out with all these details. There's probably some misunderstanding by one of us regarding how roles work, and it might well be me as I don't touch this for some time now, so let's try to find out where that is and how we can best address the problem.

As far as I understand, the fact we cannot create users with roles in arbitrary databases in any database other than admin is in fact a property of MongoDB itself, apparently confirmed by the [documentation](https://docs.mongodb.org/manual/core/security-user-defined-roles/):

>Except for roles created in the admin database, a role can only include privileges that apply to its database and can only inherit from other roles in its database.

And this seems to make sense, right?  Otherwise being able to add users to one database gives you root on the whole system.

Am I missing something here?

**Comment by mhurne** [comment]
That's a great point, and I may well be misunderstanding something here as well. Despite the documentation, I found that I was able to read data from the `bar` database using the `johndoe` example user that is in the `foo` database as described above. To prove that the `read` role assigned to that user for the `bar` database is in fact being applied I also tried a write, which failed:

```
> use foo
switched to db foo
> db.auth("johndoe","...")
1
> use bar
switched to db bar
> show collections
Foo
system.indexes
> db.Foo.find()
{ "_id" : ObjectId("56d6702b9d4757270402872d"), "foo" : "bar" }
> db.Foo.insert({ moo: "goo" })
WriteResult({
	"writeError" : {
		"code" : 13,
		"errmsg" : "not authorized on bar to execute command { insert: \"Foo\", documents: [ { _id: ObjectId('56d670d6846807b65a14b052'), moo: \"goo\" } ], ordered: true }"
	}
})
```

**Comment by mhurne** [comment]
Ahh, the key point in the documentation that you linked to is that it is talking about _roles_, not _users_.

**Comment by niemeyer** [comment]
Yes, it works because the user was created in the admin database, though, right?

**Comment by mhurne** [comment]
No, the user was created in the `foo` database (see the "db" property of the user in the output of `db.system.users.find({ user: "johndoe" }).pretty()` above).

I believe I was able to create a user in `foo` with permissions on `bar` is because I was authenticated as a user with the `root` role in the `admin` db when creating that user. But it is apparently perfectly valid for a user to exist in one DB and have permissions on some other DB.

**Comment by niemeyer** [comment]
Yes, that's the admin database, right?

```
> use admin
switched to db admin
> db.system.users.find({ user: "johndoe" }).pretty()
```

More details from the [docs](https://docs.mongodb.org/manual/reference/system-users-collection/):

> The system.users collection in the admin database stores user authentication and authorization information.

**Comment by mhurne** [comment]
I could also create such a user if I was authenticated as a user with the `userAdminAnyDatabase` role in the `admin` database when creating the user.

**Comment by mhurne** [comment]
Right, so...all user documents live in the `system.users` collection in the `admin` database, but which database the user is in logically is controlled by the `db` property on the user document. Confusing, for sure.

**Comment by mhurne** [comment]
{
    "_id" : "foo.johndoe",
    "user" : "johndoe",
    **"db" : "foo",**
   ...
}

**Comment by niemeyer** [comment]
It depends on which version of MongoDB you're sitting under. Let me clarify:

You can only have roles assigned to arbitrary databases inside the admin database. This is the case in any version of MongoDB, and it makes sense if you consider it'd be absurd to allow users in a database to arbitrate the access of users on a different database.

In MongoDB <=2.4, you could actually have users inside any database, for real. The user data was stored there. From the [docs](https://docs.mongodb.org/v2.4/reference/privilege-documents/#<database>.system.users):

> Documents in the <database>.system.users collection stores credentials and user roles for users who have access to the database.

In MongoDB >=2.6, the situation changed again, so now users are actually added to the admin database, no matter where you add them. Again, from the [docs](https://docs.mongodb.org/manual/reference/system-users-collection/):

> The system.users collection in the admin database stores user authentication and authorization information. To manage data in this collection, MongoDB provides user management commands.

So, we can probably lift the restriction when using the new commands so it's more convenient, but the driver is not actually doing much wrong. The users are still going into the admin database.

**Comment by mhurne** [comment]
> You can only have roles assigned to arbitrary databases inside the admin database. This is the case in any version of MongoDB, and it makes sense if you consider it'd be absurd to allow users in a database to arbitrate the access of users on a different database.

I understand, but I'm focused on management of _users_, not _roles_. So I don't think it is all that relevant.

I also understand what you're saying about where users are _physically_ stored in Mongo <=2.4 vs. Mongo >=2.6. But I actually don't think that's all that relevant either. mgo uses the user management commands provided by Mongo rather than dealing with the documents directly (as it should). Where Mongo decides to physically store the users is up to Mongo, and mgo doesn't need to know.

The primary issue I encountered was an inability to assign a role to a user that I should have been able to assign, and am indeed able to assign when using the mongo shell. I believe this is the relevant chunk of code:

```
if len(user.OtherDBRoles) > 0 && db.Name != "admin" && db.Name != "$external" {
		return fmt.Errorf("user with OtherDBRoles is only supported in the admin or $external databases")
}
```

The statement that "user with OtherDBRoles is only supported in the admin or $external databases" might be true in Mongo 2.4. In fact, `OtherDBRoles` is only support _in general_ in Mongo 2.4. My initial thought was that I should just avoid using `OtherDBRoles` since I'm using Mongo 3.0. But it's the only mechanism provided by mgo to assign roles to a user that aren't associated with the database that logically owns the user being created. So I _have_ to use it, but I can't unless the logical owner of the user being created is the `admin` database...

A possible solution would be to change the type of the `Roles` field on `User` from `[]Role` to `map[string]Role`. Being able to specify roles without specifying the associated database is a very minor convenience I think we could do without. mgo could continue to treat `OtherDBRoles` the way it does currently, but it should only be used when the program is running against Mongo 2.4.

**Comment by niemeyer** [comment]
> I understand, but I'm focused on management of users, not roles. So I don't think that statement is relevant.

You are talking about the management of roles inside users, and my last comment here was about exactly that.

> I also understand what you're saying about where users are physically stored in Mongo <=2.4 vs. Mongo >=2.6, but I don't think that is relevant either.

That's the whole point this issue sits on. The error check is what you see in your report in the first message, and it exists because this is a real constraint. Rather than pretending this is not relevant, we should find a way to make it more pleasing to use in the latest releases, if there's an issue about that.

> My initial thought was that I should just avoid using OtherDBRoles since I'm using Mongo 3.0. However, it's the only mechanism provided by mgo to assign roles to a user that aren't associated with the database that will logically own the created user.

Yes, the non-obvious field name is unfortunate, but it's kind of a miracle that we managed to keep a consistent API across the back and forth that happened in this area over the last several MongoDB releases.

> So I have to use it, but I can't unless the logical owner of the created user will be the admin database. But I want to use it when the owner of the created user will be some database other than admin...so I'm out of luck.

Once more: the commands you run in the shell above end up with "logical owner" of the created user being the admin database. The shell is simply pretending to you that you can put the user elsewhere, but that's not where the user lands after MongoDB 2.4. It's _always_ the admin database. Note how you _switched to the admin database_ to find the user, in your original example in this thread.

The only thing to fix in mgo is to allow it to run the command anywhere, so that it can let the server pretend to you that you're adding the user elsewhere but to the admin database. It would be fine to do that assuming that on the way we manage to not break applications running on 2.4 and under.

**Comment by mhurne** [comment]
Alright, I'm going to just use `Database.Run`. Thanks for your time discussing this!

**Comment by niemeyer** [comment]
Yeah, that'll work. I'll still do the described change, though. So let's please keep it open.

**Comment by GoranP** [comment]
I have some similar problems with Database.Run() or Session.Run() commands. I am using Mongo  3.4.9
Essentialy I want to verify if some user is in specific role. I don't want to try action on DB that will violate role.

Eg:
I created role like this in mongo shell:

`db.createRole(
   {
     role: "deploy_development",
     privileges: [       
       { resource: { db: "mydatabase", collection: "deploy_development_log" }, actions: [ "insert" ] }
     ],
    roles: []        
    }
)
`

So every user that is in this role can insert deployment data.

When I exec following command in shell.
`db.runCommand({ usersInfo: "someuser" })`

got this:
```
{
	"users" : [
		{
			"_id" : "mydatabase.someuser",
			"user" : "someuser",
			"db" : "kronocell",
			"customData" : {			
			},
			"roles" : [
				{
					"role" : "deploy_development",
					"db" : "mydatabase"
				}
			]
		}
	],
	"ok" : 1
}
```

If I try with following:
```
type MGOUsers struct {
	Users []mgo.User
}
	users := MGOUsers{}
	err := userdb.Run(bson.M{"usersInfo": "someuser"}, &users)

```
I dont get roles back. Username and custom data are binded OK.

Problem is that in _mgo.v2/session.go_ type Role is just string not the struct that server returns. {role:"", db:""}

`type Role string`


But when I create my custom structs like this:
`type MGOUsers struct {
	Users []MGOUser
}
type MGOUser struct {
	User       string
	DB         string
	CustomData interface{}
	Roles      []MGORole
}
type MGORole struct {
	Role string
	Db   string
}`

Everything is working like a charm.
Will you refactor this or is this a bigger issue that I don't see?
This type Role as a string is probably for backward compatibility. Not sure... 

Before I make Pull request I wanted your opinion on this issue. @niemeyer 

Is there any other way to get users roles?
